File: src/main/storage/repos/filters.ts
import { eq } from 'drizzle-orm';
import type { TreeVisibility } from '../../../shared/domain/tree-filter';
import { log } from '../../log';
import type { KiraDb } from '../db';
import { connectionTreeFilters } from '../schema/connection-tree-filters';

export async function listVisibility(db: KiraDb, connectionId: string): Promise<TreeVisibility> {
  const rows = await db
    .select()
    .from(connectionTreeFilters)
    .where(eq(connectionTreeFilters.connectionId, connectionId));
  const hiddenKinds: TreeVisibility['hiddenKinds'] = [];
  const hiddenPaths: string[] = [];
  for (const row of rows) {
    if (row.scope === 'kind') {
      hiddenKinds.push(row.value as TreeVisibility['hiddenKinds'][number]);
    } else if (row.scope === 'path') {
      hiddenPaths.push(row.value);
    } else {
      log('warn', 'storage/filters', `dropping unrecognised filter scope: ${row.scope}`);
    }
  }
  return { hiddenKinds, hiddenPaths };
}

// Delete-all + insert, one transaction: the dialog edits a whole set and saves it, and the set is
// small enough that a per-row diff would add complexity for no benefit (matches the old rule
// table's own replaceFilters()).
export async function replaceVisibility(
  db: KiraDb,
  connectionId: string,
  visibility: TreeVisibility,
): Promise<TreeVisibility> {
  await db.transaction(async (tx) => {
    await tx
      .delete(connectionTreeFilters)
      .where(eq(connectionTreeFilters.connectionId, connectionId));
    for (const kind of visibility.hiddenKinds) {
      await tx.insert(connectionTreeFilters).values({ connectionId, scope: 'kind', value: kind });
    }
    for (const path of visibility.hiddenPaths) {
      await tx.insert(connectionTreeFilters).values({ connectionId, scope: 'path', value: path });
    }
  });
  return listVisibility(db, connectionId);
}

File: apps/kira-studio/internal/storage/repos/filters.go
package repos

import (
	"database/sql"
	"fmt"
	"log/slog"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

type FiltersRepo struct {
	DB *sql.DB
}

func (r *FiltersRepo) List(connectionID string) (model.TreeVisibility, error) {
	rows, err := r.DB.Query(
		`SELECT scope, value FROM connection_tree_filters WHERE connection_id = ?`, connectionID,
	)
	if err != nil {
		return model.TreeVisibility{}, fmt.Errorf("repos/filters: query: %w", err)
	}
	defer rows.Close()

	out := model.EmptyVisibility()
	for rows.Next() {
		var scope, value string
		if err := rows.Scan(&scope, &value); err != nil {
			return model.TreeVisibility{}, fmt.Errorf("repos/filters: scan: %w", err)
		}
		switch scope {
		case "kind":
			out.HiddenKinds = append(out.HiddenKinds, value)
		case "path":
			out.HiddenPaths = append(out.HiddenPaths, value)
		default:
			slog.Warn("dropping unrecognised filter scope", "scope", "storage/filters", "connectionId", connectionID, "rowScope", scope)
		}
	}
	if err := rows.Err(); err != nil {
		return model.TreeVisibility{}, fmt.Errorf("repos/filters: rows: %w", err)
	}
	return out, nil
}

// Replace is delete-all + insert, one transaction (filters.ts's replaceVisibility): the dialog
// edits a whole set and saves it. INSERT OR IGNORE (D10) absorbs a duplicate value in the input —
// the primary key is (connection_id, scope, value) and the payload is a set, so a repeat is a
// no-op rather than an error.
func (r *FiltersRepo) Replace(connectionID string, v model.TreeVisibility) (model.TreeVisibility, error) {
	tx, err := r.DB.Begin()
	if err != nil {
		return model.TreeVisibility{}, fmt.Errorf("repos/filters: begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	if _, err := tx.Exec(`DELETE FROM connection_tree_filters WHERE connection_id = ?`, connectionID); err != nil {
		return model.TreeVisibility{}, fmt.Errorf("repos/filters: clear: %w", err)
	}
	for _, kind := range v.HiddenKinds {
		if _, err := tx.Exec(
			`INSERT OR IGNORE INTO connection_tree_filters (connection_id, scope, value) VALUES (?, 'kind', ?)`,
			connectionID, kind,
		); err != nil {
			return model.TreeVisibility{}, fmt.Errorf("repos/filters: insert kind %s: %w", kind, err)
		}
	}
	for _, path := range v.HiddenPaths {
		if _, err := tx.Exec(
			`INSERT OR IGNORE INTO connection_tree_filters (connection_id, scope, value) VALUES (?, 'path', ?)`,
			connectionID, path,
		); err != nil {
			return model.TreeVisibility{}, fmt.Errorf("repos/filters: insert path %s: %w", path, err)
		}
	}
	if err := tx.Commit(); err != nil {
		return model.TreeVisibility{}, fmt.Errorf("repos/filters: commit: %w", err)
	}
	return r.List(connectionID)
}
